demos/http3/ossl-nghttp3-demo-server.c

```diff
@@ -0,0 +1,640 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+#include <assert.h>
+#include <netinet/in.h>
+#include <nghttp3/nghttp3.h>
+#include <openssl/err.h>
+#include <openssl/quic.h>
+#include <openssl/ssl.h>
+#include <unistd.h>
+
+#define MAKE_NV(NAME, VALUE)                                                   \
+  {(uint8_t *)(NAME), (uint8_t *)(VALUE), sizeof((NAME)) - 1,                  \
+   sizeof((VALUE)) - 1, NGHTTP3_NV_FLAG_NONE}
+#define nghttp3_arraylen(A) (sizeof(A) / sizeof(*(A)))
+
+/* CURL according to trace has 2 more streams 7 and 11 */
+struct ssl_id {
+  SSL *s;
+  uint64_t id;
+};
+
+#define MAXSSL_IDS 20
+struct h3ssl {
+  struct ssl_id ssl_ids[MAXSSL_IDS];
+  int end_headers_received;
+  int datadone;
+  int has_uni;
+  int done;
+};
+
+static void init_id(struct h3ssl *h3ssl) {
+  struct ssl_id *ssl_ids;
+  ssl_ids = h3ssl->ssl_ids;
+  for (int i = 0; i < MAXSSL_IDS; i++) {
+    ssl_ids[i].s = NULL;
+    ssl_ids[i].id = -1;
+  }
+  h3ssl->end_headers_received = 0;
+  h3ssl->datadone = 0;
+  h3ssl->has_uni = 0;
+  h3ssl->done = 0;
+}
+
+static void add_id(uint64_t id, SSL *ssl, struct h3ssl *h3ssl) {
+  struct ssl_id *ssl_ids;
+  ssl_ids = h3ssl->ssl_ids;
+  for (int i = 0; i < MAXSSL_IDS; i++) {
+    if (!ssl_ids[i].s) {
+      ssl_ids[i].s = ssl;
+      ssl_ids[i].id = id;
+      return;
+    }
+  }
+  printf("Oops too many streams to add!!!\n");
+  exit(1);
+}
+
+static void h3close(struct h3ssl *h3ssl, uint64_t id) {
+  struct ssl_id *ssl_ids;
+  ssl_ids = h3ssl->ssl_ids;
+  for (int i = 0; i < MAXSSL_IDS; i++) {
+    if (ssl_ids[i].id == id) {
+      if (!SSL_stream_conclude(ssl_ids[i].s, 0))
+        goto err;
+      err:
+      SSL_shutdown(ssl_ids[i].s);
```

Comment:
@jfclere I'm confused by this part. The `if` statement seems useless given the error label gets executed anyway. Am I missing something?

Reply:
SSL_stream_conclude() return code needs to be tested, should I put a warning there?

Reply:
I don't have any issue with the test itself, I just don't understand why the `goto` is there when the `SSL_shutdown` call gets called anyway. Now, the code is equivalent to

```.c
!SSL_stream_conclude(ssl_ids[i].s, 0);
SSL_shutdown(ssl_ids[i].s);
```

which seems like a bug.

Reply:
Meaning that the result of the test has no meaningful effect on the code behavior. Or I'm terribly missing something..

Reply:
Correct I should print a message.